From ac1b9f4b4059c28cca47417425344a62d5dd197f Mon Sep 17 00:00:00 2001
From: Kyunglyul Hyun <klhyun@google.com>
Date: Wed, 2 Oct 2024 03:50:00 +0000
Subject: [PATCH] Reset permissions for not bonded device

According to the PBAP specification,
The PSE user shall have to confirm at least the first Phone Book Access
Profile connection from each new PCE.

According to the MAP specification,
The MCE and MSE shall be bonded before setting up a Message Access Profile
connection.

Let's remove the permissions when the device is unbonded.

Flag: EXEMPT, security fix
Bug: 289375038
Bug: 289811388
Test: atest BluetoothInstrumentationTests
Ignore-AOSP-First: security fix

(cherry picked from commit 771aed3ea225d9ca2ef91e88f45f330ced952772)
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:8ef714b803127ca193773c6d032457c01c1963d6)
Merged-In: I94478a4a45a434732d4bc4810da31ef45ef891af
Change-Id: I94478a4a45a434732d4bc4810da31ef45ef891af
---
 .../com/android/bluetooth/btservice/AdapterService.java   | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/android/app/src/com/android/bluetooth/btservice/AdapterService.java b/android/app/src/com/android/bluetooth/btservice/AdapterService.java
index 62a532dccd..70408cc666 100644
--- a/android/app/src/com/android/bluetooth/btservice/AdapterService.java
+++ b/android/app/src/com/android/bluetooth/btservice/AdapterService.java
@@ -27,6 +27,7 @@ import static android.bluetooth.BluetoothAdapter.SCAN_MODE_CONNECTABLE;
 import static android.bluetooth.BluetoothAdapter.SCAN_MODE_CONNECTABLE_DISCOVERABLE;
 import static android.bluetooth.BluetoothAdapter.SCAN_MODE_NONE;
 import static android.bluetooth.BluetoothDevice.BATTERY_LEVEL_UNKNOWN;
+import static android.bluetooth.BluetoothDevice.BOND_NONE;
 import static android.bluetooth.BluetoothDevice.TRANSPORT_AUTO;
 import static android.bluetooth.IBluetoothLeAudio.LE_AUDIO_GROUP_ID_INVALID;
 import static android.text.format.DateUtils.MINUTE_IN_MILLIS;
@@ -6109,6 +6110,13 @@ public class AdapterService extends Service {
             mCsipSetCoordinatorService.handleBondStateChanged(device, fromState, toState);
         }
         mDatabaseManager.handleBondStateChanged(device, fromState, toState);
+
+        if (toState == BOND_NONE) {
+            // Remove the permissions for unbonded devices
+            setMessageAccessPermission(device, BluetoothDevice.ACCESS_UNKNOWN);
+            setPhonebookAccessPermission(device, BluetoothDevice.ACCESS_UNKNOWN);
+            setSimAccessPermission(device, BluetoothDevice.ACCESS_UNKNOWN);
+        }
     }
 
     static int convertScanModeToHal(int mode) {
-- 
2.34.1

